ext/web/08_text_encoding.js

```diff
@@ -96,52 +96,68 @@
       });
 
       // TODO(lucacasonato): add fast path for non-streaming decoder & decode
-
+      /*
       if (this.#rid === null) {
         this.#rid = core.opSync("op_encoding_new_decoder", {
           label: this.#encoding,
           fatal: this.#fatal,
           ignoreBom: this.#ignoreBOM,
         });
-      }
+      }*/
 
+      //try {
       try {
-        try {
-          if (ArrayBufferIsView(input)) {
-            input = new Uint8Array(
-              input.buffer,
-              input.byteOffset,
-              input.byteLength,
-            );
-          } else {
-            input = new Uint8Array(input);
-          }
-        } catch {
-          // If the buffer is detached, just create a new empty Uint8Array.
-          input = new Uint8Array();
-        }
-        if (
-          ObjectPrototypeIsPrototypeOf(
-            SharedArrayBuffer.prototype,
+        if (ArrayBufferIsView(input)) {
+          input = new Uint8Array(
             input.buffer,
-          )
-        ) {
-          // We clone the data into a non-shared ArrayBuffer so we can pass it
-          // to Rust.
-          // `input` is now a Uint8Array, and calling the TypedArray constructor
-          // with a TypedArray argument copies the data.
+            input.byteOffset,
+            input.byteLength,
+          );
+        } else {
           input = new Uint8Array(input);
         }
-        return core.opSync("op_encoding_decode", input, {
-          rid: this.#rid,
-          stream: options.stream,
+      } catch {
+        // If the buffer is detached, just create a new empty Uint8Array.
+        input = new Uint8Array();
+      }
+      if (
+        ObjectPrototypeIsPrototypeOf(
+          SharedArrayBuffer.prototype,
+          input.buffer,
+        )
+      ) {
+        // We clone the data into a non-shared ArrayBuffer so we can pass it
+        // to Rust.
+        // `input` is now a Uint8Array, and calling the TypedArray constructor
+        // with a TypedArray argument copies the data.
+        input = new Uint8Array(input);
+      }
+      
+      if (!options.stream) {
```

Comment:
```suggestion
      if (!options.stream && this.#rid === null) {
```

`stream: true` makes the current decoding context (i.e. the current `Decoder` instance) be kept around for the next call to `TextDecoder.prototype.decode`, rather than end with this call. Using `op_encoding_decode_single` when there is an open decoding context means that any state kept from the last call (for example, bytes at the end of the last input that couldn't be decoded without having access to the remainder of the input) would be discarded.

```js
const decoder = new TextDecoder();

// Without `{stream: true}` this would return "🙈�", because the last two bytes
// aren't valid UTF-8 on their own.
decoder.decode(
  new Uint8Array([0xF0, 0x9F, 0x99, 0x88, 0xF0, 0x9F]),
  {stream: true}
);  // "🙈"

// With your PR as it is now, this would return "�🙊", because the first two
// bytes aren't valid UTF-8 on their own, but they are when appended to the
// last input.
decoder.decode(new Uint8Array([0x99, 0x89, 0xF0, 0x9F, 0x99, 0x8A])); // "🙉🙊"
```

With this change, the assumption that `options.stream` can't be false in the code below no longer holds.

Reply:
Thank you @andreubotella! 
The failed tests were reduced to 6 now.  Will have a closer look at the remaining failed tests and your comments in a few hours.

Reply:
All tests are passing now. The first version of this refactoring is complete. Please have a look.